pages/security/csp.md

```diff
@@ -0,0 +1,115 @@
+
+To do this, use the `report-uri` directive, passing it a fully qualified URI e.g. `https://my-public-site.gov/reports/csp`. Now, each time content from an invalid source is encountered, your browser will POST a JSON payload to the provided URI.
+
+Here is an example, from the MDN article linked in [Further Reading](#further-reader)
+
+```
+{
+  "csp-report": {
+    "document-uri": "http://example.com/signup.html",
+    "referrer": "",
+    "blocked-uri": "http://example.com/css/style.css",
+    "violated-directive": "style-src cdn.example.com",
+    "original-policy": "default-src 'none'; style-src cdn.example.com; report-uri /_/csp-reports"
+  }
+}
+```
+
+This data can of course be mixed with other request metadata, like IP address or browser User-Agent, and fed into whatever log processing system your app has in place.
+
+Reporting can only be enabled via HTTP header, not inside a `<meta>` tag!
+
+
+
+### Client-side Implementation
+To implement CSP on the client, add a `<meta>` tag to your web site's `<head>`.
+
+`<meta http-equiv="Content-Security-Policy" content="default-src 'self'; img-src https://*; child-src 'none';">`
+
+You're done!
+
+### Server-side Implementation
+Similarly, to include CSP on the server, simply return a `Content-Security-Policy: {my-policy-string}` HTTP header in your web responses.
+
+<br>
+
+Although a basic CSP policy is relatively easy to write, you might find it easier to use a server-side library to define your policies declaratively. Two of well-known and actively maintained libraries are [secureheaders](https://github.com/twitter/secureheaders), a **ruby** gem written and maintained by Twitter, and the [helmetjs](https://github.com/helmetjs/helmet) middleware for **express.js**.
+
+## Caveats
+A major caveat is that CSP, by default, prevents developers from adding inline `<style>` and `<script>` tags, and from specifying JavaScript behavior or CSS `style` attributes directly on DOM elements. This isn't a bad thing, as current best practices in front-end development discourage the use of these patterns.
+
+However, there are scenarios in which you may want to use an inline content tag, such as including a Google Analytics initialization script. For these cases, CSP allows the user to supply a `nonce` attribute to the inline script: `<script nonce=GHDGsfsd83hfdfFD3>...javascript here...</script>`. Then, in your policy, `script-src 'nonce-GHDGsfsd83hfdfFD3'`. Nonces must be regenerated on each request. Additionally, CSP 2.0 allows you to generate a SHA hash of the script content, and pass that to the `srcipt-src` directive as `'sha-{content-sha}'`. You don't need to use a `nonce` attribute if your content is hashed.
+
+Another potential issue is third-party libraries that automatically inject JavaScript and CSS into your HTML. If your project utilizes a library that does this, the only solution is
```

Comment:
This isn't strictly true, technically one could make a SHA256 hash of the contents of the injected javascript, and whitelist that. I didn't mention it because it seems really hacky and error-prone, as the hash would have to be recomputed every time the script changed.

That being said it is a valid method, and someone might decide it's an acceptable tradeoff, so maybe it's worth mentioning?

Reply:
I'm fine with mentioning it with caveats!